### ISSUE TITLE
Optimize client queue to use a circular buffer

### ISSUE THREAD

**Issue by ronag** [nodejs/undici#355]
`client[kQueue]` could be implemented as a circular queue of `size >= client[kPipelining]` which would remove the need to splice it.

**Comment by dnlup** [comment]
You are referring to the `FixedQueue` implemented in https://github.com/nodejs/undici/blob/master/lib/core/node/fixed-queue.js? This is the queue that the client should use?

**Comment by ronag** [comment]
>  This is the queue that the client should use?

No that's unfortunately not possible since the queue is partitioned. Just a regular circular queue should work since the queue size should preferably not be larger than the pipelining factor.

**Comment by dnlup** [comment]
I made a POC trying to implement the circular queue you suggested [here](https://github.com/dnlup/seeq/blob/next/index.js). Does it come near to what you had in mind? I wanted to be sure I understood before trying to put it in `undici`. The idea is that instead of using `splice,` you iterate over it directly. You use `push` and `pop` for the other operations. Maybe I didn' figure out all the details yet. Performance-wise, it is slightly more performant.

**Comment by mcollina** [comment]
Did you run any benchmarks if this would improve undici?

**Comment by dnlup** [comment]
No, I didn't. I'll try to embed it in undici and run them. There's a bench of the queue vs array.splice [here](https://github.com/dnlup/seeq/blob/next/bench.js) that gives [these results](https://github.com/dnlup/seeq/blob/next/bench.txt) on my machine. I don't know how and if it would impact undici yet.

**Comment by ronag** [comment]
A more realistic size would be ~16.

**Comment by dnlup** [comment]
You're right @ronag , using `16` the improvement is even more marginal:

https://github.com/dnlup/seeq/blob/next/bench.txt

**Comment by dnlup** [comment]
Ops, I also forgot about the fact the `pipelining` can be changed at runtime:

* https://github.com/nodejs/undici/blob/master/lib/core/client.js#L182

I should add this feature to the circular queue too.

**Comment by dnlup** [comment]
While refactoring for this feature, I changed a bit the benchmarks, and it looks like my implementation is actually slower than using `slice`  :

https://github.com/dnlup/seeq/blob/next/bench.txt

I'll continue doing some tests, but maybe the queue I linked doesn't improve anything to start with. Sorry for the noise.

**Comment by ronag** [comment]
Let me know what you find and then maybe we can close this issue either with an update or as a wontfix.

**Comment by dnlup** [comment]
Quick update. I thnk I managed to write an implementation that performs better using a circular linked list instead of an Array:

https://github.com/dnlup/seeq/blob/next/bench.txt

In the benchmark, I am not recreating the queue each time because I think it was the main reason to try this approach instead of using `Array.prototype.slice`. I am creating an array though, hoping that this would highlight just the difference of speed in the manipulation of the elements.

I am still not sure I got the API right. I am trying try to embed it in undici to see what happens. For now, I am still in the phase where it explodes :boom: . I have to get a better grasp of how the queue was working before and how/if the new one will fit.
